linkerd/app/core/src/http_tracing.rs

```diff
@@ -1,139 +1,88 @@
+use crate::http_tracing::opencensus::{OcSpanConverter, OpenCensusSink};
+use crate::http_tracing::opentelemetry::{OpenTelemetrySink, OtelSpanConverter};
 use linkerd_error::Error;
-use linkerd_opencensus::proto::trace::v1 as oc;
 use linkerd_stack::layer;
-use linkerd_trace_context::{self as trace_context, TraceContext};
+use linkerd_trace_context as trace_context;
+use linkerd_trace_context::{Span, TraceContext};
 use std::{collections::HashMap, sync::Arc};
 use thiserror::Error;
-use tokio::sync::mpsc;
 
-pub type OpenCensusSink = Option<mpsc::Sender<oc::Span>>;
+mod opencensus;
+mod opentelemetry;
+
 pub type Labels = Arc<HashMap<String, String>>;
 
-/// SpanConverter converts trace_context::Span objects into OpenCensus agent
-/// protobuf span objects. SpanConverter receives trace_context::Span objects by
-/// implmenting the SpanSink trait. For each span that it receives, it converts
-/// it to an OpenCensus span and then sends it on the provided mpsc::Sender.
-#[derive(Clone)]
-pub struct SpanConverter {
-    kind: Kind,
-    sink: mpsc::Sender<oc::Span>,
-    labels: Labels,
+#[derive(Clone, Debug)]
+pub enum SpanSink {
+    Oc(OpenCensusSink),
+    Otel(OpenTelemetrySink),
```

Comment:
I'm a little surprised that we need to expose the type of the tracing backend to the span producer. I was expecting that we'd have a common span type from the publisher's point of view and that the converter would handle converting the span to the proper wire format. For example, is it possible to use the SpanData type in both cases?

Reply:
I agree it's a leaky abstraction that ideally shouldn't exist. I spent some time digging through this, turns out it's a lot easier to re-use our existing internal span representation and send that to the exporter task, letting the exporter be the one that converts the span to the protocol-specific format. This has the added benefit of offloading the (fairly small) overhead of converting the trace format onto the exporter rather than the tracing middleware.